### PR TITLE
PaymentProcessor - Enter title instead of name on form, call hooks

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Adds a new "Frontend title" to Payment Processors for end-users, and redefines "Title" as a backend field for admins.

Before
----------------------------------------
- "Name" is required, "Title" is not.
- "Title" is more-or-less understood to be a frontend thing, while "Name" is admin-facing.

image

After
----------------------------------------
- "Name" is hidden from the form completely and auto-calculated from "Title" (this is an established pattern for many entities).
- "Title" is explicitly described as backend for admin eyes only.
- "Frontend Title" field added, which is explicitly end-user-facing.

image

Technical Details
----------------------------------------
- Updates the PaymentProcessor form/dao to use hooks instead of an overloaded "create" function
- This necessitated an additional argument to `hook_civicrm_post()` which seems generally useful.
- The upgrade step populates "Frontend Title" with `COALESCE(title, name)` because "Title" is sometimes blank.
- The upgrade step populates "Title" with "Name" to reflect the new meaning of those fields.
- "Name" will be autogenerated from "Title" in the `DAO::writeRecord` function.
- "Title" and "Frontend Title" are both required in the DB, but the API is forgiving and will use name as title and title as frontend title if values are omitted.

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
@colemanw the reason I keep mentioning frontend_title is that 'name' is defaco the back-office tittle eg
```
'name' => 'Eway'
'title' => 'Credit card'
```

Eway shows to back office users. People filling in the form see 'Credit card' (cos they don't care who the gateway is)

**Comment by colemanw** [comment]
@eileenmcnaughton this PR might be large enough already before it gets unreviewable so maybe add that column as a followup?

**Comment by eileenmcnaughton** [comment]
@colemanw I'm OK to review this PR without that (once it passes) - but it would be experienced as a regression if we 'take away' name field without putting in place the title/frontend_title to meet the needs name/title are currently meeting

File: CRM/Admin/Form/PaymentProcessor.php [hunk 1]
```diff
@@ -486,7 +476,9 @@ public function updatePaymentProcessor(&$values, $domainID, $test) {
       'accepted_credit_cards' => $creditCards,
     ], $values);
 
-    civicrm_api3('PaymentProcessor', 'create', $params);
+    $result = civicrm_api3('PaymentProcessor', 'create', $params);
```
Comment (eileenmcnaughton) [inline CRM/Admin/Form/PaymentProcessor.php:479]: Might as well switch to use v4 since it will get r-run testing which is the hardest part of switching :-)

Comment (colemanw) [reply CRM/Admin/Form/PaymentProcessor.php]: Ok done

**Comment by seamuslee001** [comment]
so I think there is some issues in the civicrm_data then and so we are kind of changing things when name has = backend name and title = the front end name for the payment processor right?

**Comment by mattwire** [comment]
Description on this PR needs updating to reflect what it's actually doing now?

**Comment by colemanw** [comment]
I think running `regen.sh` will fix the test site falling over. But first...
@mattwire the description does describe the PR accurately, however @eileenmcnaughton is pushing for more changes, and now that I think about it, they kind of are necessary because we're in kindof an "all or nothing" situation here.

The current state is that "name" gets used for backend and "title" for the frontend (@eileenmcnaughton is this actually implemented everywhere or is `title` still ignored in some frontend contexts).

The desired state is for "name" to be internal, "title" for the backend and "frontend_title" for the frontend.

I don't think there's a viable "in-between" state. But to get us all the way we're looking at an increased scope of:

1. Adding the new `frontend_title` column
2. An upgrade step to populate it (implementing the logic that we've now switched the meaning of "title" from frontend to backend"
3. Updating all frontend forms

**Comment by seamuslee001** [comment]
@colemanw @eileenmcnaughton  re using title on front end forms I think that already happens as this https://github.com/civicrm/civicrm-core/blob/master/CRM/Financial/Form/FrontEndPaymentFormTrait.php#L176 is called when adding processors to front end forms

**Comment by eileenmcnaughton** [comment]
@seamuslee001 & to be explicit - I think that function you link to should be updated to ONLY return `frontend_title` - so that whatever they put in there (even nothing) is respected.

**Comment by colemanw** [comment]
@eileenmcnaughton after much pain, tests are now passing. This is a big diff due to all the tests/sample data that needed updating and it is prone to going stale.
I've rewritten the PR description to reflect what's now happening. Over to you for "Updating all frontend forms"

**Comment by eileenmcnaughton** [comment]
@colemanw I'm still looking but so far

1) the upgrade script worked
2) I feel like we said frontend_title would be required but that doesn't seem to be the case - I can't quite recall where we got to here
3) When i used the form to update an existing processor the 'name' field was replaced with the word 'null'

**Comment by seamuslee001** [comment]
Also why is there no update the frontEndFormTrait?

File: CRM/Financial/BAO/PaymentProcessor.php [hunk 2]
```diff
@@ -184,20 +131,72 @@ public static function del($paymentProcessorID) {
     static::deleteRecord(['id' => $paymentProcessorID]);
   }
 
+  /**
+   * Callback for hook_civicrm_pre().
+   * @param \Civi\Core\Event\PreEvent $event
+   */
+  public static function self_hook_civicrm_pre(\Civi\Core\Event\PreEvent $event) {
+    if ($event->action === 'create') {
+      // If we are creating a new PaymentProcessor and have not specified the payment instrument to use, get the default from the Payment Processor Type.
```
Comment (eileenmcnaughton) [inline CRM/Financial/BAO/PaymentProcessor.php:140]: This was already the case - but this is actually just another parameter on the next chunk of code

Comment (colemanw) [reply CRM/Financial/BAO/PaymentProcessor.php:140]: This code is not new, it's just moved out of the `create` function

Comment (eileenmcnaughton) [reply CRM/Financial/BAO/PaymentProcessor.php:140]: Yeah - that was what I meant about 'already the case'

**Comment by eileenmcnaughton** [comment]
I'm looking at the form - I see why it's an issue

**Comment by eileenmcnaughton** [comment]
I pushed in a fix for the form messing up the save -  but it's way too easy to save with a blank frontend_title - I think making it required is the easiest here since no-one will not have one as yet

**Comment by eileenmcnaughton** [comment]
@artfulrobot yes to both

**Comment by eileenmcnaughton** [comment]
@artfulrobot note however there is still a game on `name` here - where the live has to match the test - I feel like the right fix is to add a new `master_id` column & make the names truly unique

**Comment by colemanw** [comment]
> it's way too easy to save with a blank frontend_title - I think making it required is the easiest here since no-one will not have one as yet

@eileenmcnaughton that sounds fine to make it required on the form. It's not required in the DB or by the API, and we just need to decide whether NULL means "fall back to `title`" or if it means "show it as blank to the user"

**Comment by eileenmcnaughton** [comment]
@colemanw I think if there is any fallback it should occur when creating - not when rendering. I don't think there is a use case for making it blank - you can't at the moment & no-one has requested it. Since the label doesn't show if there is only one processor enabled it isn't like profile where people just want to show no heading. So I think we should

1) make it required at the form & db level
2) make it optional at the api level, populating from title if nothing is provided

**Comment by mattwire** [comment]
I'm not sure of the concept of a "master" processor. If the names match that links them together, the title/frontend_title could be customised if desired (but I don't think we need a UI for that).

**Comment by mattwire** [comment]
> @artfulrobot @mattwire the pain with `name` is that it is not unique so breaks assumption about name being a unique thing - I guess we could achieve the same thing by changing the convention to `test_$name`

Ah, yes. Makes sense.

**Comment by eileenmcnaughton** [comment]
@colemanw this went stale too

**Comment by colemanw** [comment]
> the pain with `name` is that it is not unique so breaks assumption about name being a unique thing - I guess we could achieve the same thing by changing the convention to `test_$name`

In this case the unique index is on `name` + `is_test`, which is fine IMO I just don't care enough to put effort into changing it, but maybe @eileenmcnaughton does...

**Comment by eileenmcnaughton** [comment]
I think that v3 test fail is some accidental messing with `is_test` - will check

**Comment by eileenmcnaughton** [comment]
@colemanw looks like you force pushed out my fixes the the PaymentProcessor form - but I think I have htem locally

**Comment by colemanw** [comment]
@eileenmcnaughton I just force-pushed again to fix an APIv3 test.
Looks like gitHub still has your commit here: https://github.com/civicrm/civicrm-core/commit/b41d07876b934117614730e585e3756aedf81d6e but I don't see it in your fork so can't cherry-pick it.

**Comment by eileenmcnaughton** [comment]
@colemanw pushing now

**Comment by eileenmcnaughton** [comment]
@colemanw here - https://github.com/civicrm/civicrm-core/compare/master...eileenmcnaughton:civicrm-core:ppUpdate?expand=1 - but I'm just about to re-test

File: CRM/Financial/BAO/PaymentProcessor.php [hunk 3]
```diff
@@ -184,20 +131,79 @@ public static function del($paymentProcessorID) {
...
+        $event->params['payment_instrument_id'] = civicrm_api3('PaymentProcessorType', 'getvalue', [
+          'id' => $event->params['payment_processor_type_id'],
+          'return' => 'payment_instrument_id',
+        ]);
+      }
+      // Supply defaults for `title` and `frontend_title`
+      if (!isset($event->params['title'])) {
+        $event->params['title'] = $event->params['name'];
+      }
+      if (!isset($event->params['frontend_title'])) {
```
Comment (eileenmcnaughton) [inline CRM/Financial/BAO/PaymentProcessor.php:151]: Ah ok - this was the bit I thought was going into `writeRecord` - but yeah it doesn't have to

**Comment by eileenmcnaughton** [comment]
@colemanw I'm having problems with the upgrade script locally - I think it's cos it's trying to call `runSQL` twice and I *think* the reason you are calling that rather than just doing the sql in php is you think it handles `localise` better - I'm not quite clear on the circumstances here -  `executeQuery` handles localise - but maybe not in upgrade mode?

Does it run through for you?

**Comment by eileenmcnaughton** [comment]
@colemanw I put up a PR against this so you can look at what made the form layer work for me. I'm still not sure re the upgrade - I think there is an issue that affects it only when running through the UI per previous comment

**Comment by eileenmcnaughton** [comment]
ok - I think this might be the upgrade issue - https://github.com/civicrm/civicrm-core/pull/25940 - it was just merged - so I think just the form saving issues

**Comment by colemanw** [comment]
@eileenmcnaughton I've rebased/cherry-picked in your commits, thanks.

File: CRM/Financial/BAO/PaymentProcessor.php [hunk 4]
```diff
@@ -153,15 +149,19 @@ public static function self_hook_civicrm_pre(\Civi\Core\Event\PreEvent $event) {
...
         throw new CRM_Core_Exception(ts('Could not find payment processor meta information'));
       }
-      $event->params['is_recur'] = $ppTypeDAO->is_recur;
-      $event->params['billing_mode'] = $ppTypeDAO->billing_mode;
-      $event->params['class_name'] = $ppTypeDAO->class_name;
-      $event->params['payment_type'] = $ppTypeDAO->payment_type;
+      // If we are creating a new PaymentProcessor and have not specified the payment instrument to use, get the default from the Payment Processor Type.
+      if (empty($event->params['payment_instrument_id'])) {
+        $event->params['payment_instrument_id'] = $paymentProcessorType->payment_instrument_id;
+      }
```
Comment (colemanw) [inline CRM/Financial/BAO/PaymentProcessor.php:160]: @eileenmcnaughton I updated this commit to make this conditional and now the tests are passing.

Comment (eileenmcnaughton) [reply CRM/Financial/BAO/PaymentProcessor.php:160]: that makes sense!

**Comment by eileenmcnaughton** [comment]
I hit some issues testing the upgrade but I think it was around my db having an inconsistent state - the part related to adding the field went ok.

I re-tested the payment processor form & that worked!

I think this is good to merge & we should update the other entities with frontend_title in the same release

**Comment by larssandergreen** [comment]
Found a regression from this PR: https://lab.civicrm.org/dev/core/-/issues/4418

**Pull request merged**